### ISSUE TITLE
Denoting members by name in a library augmentation or a library that has augmentations

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#3717]
Thanks to @sgrekhov for bringing up this issue. Consider the following situation:

```dart
// --- Library augmentation 'augment.dart'.
augment library 'main.dart';

augment int foo(int i) => i > 2 ? foo(i - 1) : augmented(i - 1);

// --- Library 'main.dart'.
import augment 'augment.dart';

int foo(int i) {
  if (i > 0) return foo(i);
  return i;
}
```

It is my understanding that `foo(...)` in every location denotes an invocation the final, merged result of introducing the declaration named `foo` in `main.dart` and augmenting it in some number of steps (in this case just once). In general, a named reference to an instance member is always the post-merge member (the "end result"), never a single augmentation, not even when it occurs in the original (non-augmenting) declaration (here: the declaration in `main.dart`).

In particular, the invocation of `foo(i)` in the original declaration is not an infinite loop.

I couldn't find an explicit answer to this question in the [feature specification](https://github.com/dart-lang/language/blob/main/working/augmentation-libraries/feature-specification.md). It is possible that it is considered obvious, but it is still useful to get it clarified, and we might also add a couple of sentences to the feature specification about it.

**Comment by jakemac53** [comment]
Agreed, `augmented` is the only way to refer to anything other than the final merged member.

**Comment by eernstg** [comment]
I'll conclude that the name of a member `foo` in an original or augmenting declaration body, when it occurs in a declaration _D_ named `foo`, or elsewhere such that `foo` denotes _D_, always denotes the final result of performing the augmentations. The only way each of the augmenting declarations can be denoted is as `augmented` in an augmenting declaration.

**Comment by davidmorgan** [comment]
@eernstg looks like this can be closed if #3874 looks good to you? :)

**Comment by eernstg** [comment]
#3874 does look good, closing this one, thanks!
